hmpps-sqs-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/sqs/HmppsAsyncQueueFactory.kt

```diff
@@ -0,0 +1,121 @@
+          .also { log.info("Created ${hmppsSqsProperties.provider} SqsAsyncClient for DLQ queueId $queueId with name ${queueConfig.dlqName}") }
+      }
+    } else null
+
+  private fun getOrDefaultSqsAsyncClient(queueId: String, queueConfig: QueueConfig, hmppsSqsProperties: HmppsSqsProperties, sqsDlqClient: SqsAsyncClient?): SqsAsyncClient =
+    getOrDefaultBean("$queueId-sqs-client") {
+      createSqsAsyncClient(queueConfig, hmppsSqsProperties, sqsDlqClient)
+        .also { log.info("Created ${hmppsSqsProperties.provider} SqsAsyncClient for queue queueId $queueId with name ${queueConfig.queueName}") }
+    }
+
+  private fun getOrDefaultAsyncHealthIndicator(hmppsQueue: HmppsAsyncQueue): HealthIndicator =
+    getOrDefaultBean("${hmppsQueue.id}-health") {
+      HmppsAsyncQueueHealth(hmppsQueue)
+    }
+
+  @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+  private inline fun <reified T> getOrDefaultBean(beanName: String, createDefaultBean: () -> T) =
+    runCatching { context.beanFactory.getBean(beanName) as T }
+      .getOrElse {
+        createDefaultBean().also { bean -> context.beanFactory.registerSingleton(beanName, bean) }
+      }
+
+  fun createSqsAsyncDlqClient(queueConfig: QueueConfig, hmppsSqsProperties: HmppsSqsProperties): SqsAsyncClient {
+    val region = hmppsSqsProperties.region
+    val provider = findProvider(hmppsSqsProperties.provider)
+    if (queueConfig.dlqName.isEmpty()) throw MissingDlqNameException()
+    return when (provider) {
+      Provider.AWS -> sqsAsyncClientFactory.awsSqsAsyncClient(queueConfig.dlqAccessKeyId, queueConfig.dlqSecretAccessKey, region)
+      Provider.LOCALSTACK -> sqsAsyncClientFactory.localstackSqsAsyncClient(hmppsSqsProperties.localstackUrl, region)
+        .also { sqsDlqClient -> sqsDlqClient.createQueue(CreateQueueRequest.builder().queueName(queueConfig.dlqName).build()) }
+    }
+  }
+
+  fun createSqsAsyncClient(queueConfig: QueueConfig, hmppsSqsProperties: HmppsSqsProperties, sqsDlqClient: SqsAsyncClient?): SqsAsyncClient {
+    val region = hmppsSqsProperties.region
+    return when (findProvider(hmppsSqsProperties.provider)) {
+      Provider.AWS -> sqsAsyncClientFactory.awsSqsAsyncClient(queueConfig.queueAccessKeyId, queueConfig.queueSecretAccessKey, region)
+      Provider.LOCALSTACK -> sqsAsyncClientFactory.localstackSqsAsyncClient(hmppsSqsProperties.localstackUrl, region)
+        .also { sqsClient -> createLocalStackQueueAsync(sqsClient, sqsDlqClient, queueConfig.queueName, queueConfig.dlqName) }
+    }
+  }
+
+  private fun createLocalStackQueueAsync(
+    sqsClient: SqsAsyncClient,
+    sqsDlqClient: SqsAsyncClient?,
+    queueName: String,
+    dlqName: String,
+  ) {
+    if (dlqName.isEmpty() || sqsDlqClient == null) {
+      sqsClient.createQueue(CreateQueueRequest.builder().queueName(queueName).build())
+    } else {
+      sqsDlqClient.getQueueUrl(GetQueueUrlRequest.builder().queueName(dlqName).build())
```

Comment:
I'm guessing this code is a candidate for sharing between non-async version - I guess it is identical